File: Server~/src/tools/physics/RaycastTool.ts
import { z } from 'zod';
import { BaseTool } from '../base/BaseTool.js';
import { Tool, Tags, Examples } from '../base/ToolDecorators.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const QueryTriggerInteractionSchema = z.enum(['UseGlobal', 'Ignore', 'Collide']);

const RaycastArgsSchema = z.object({
  origin: z.object({ x: z.number(), y: z.number(), z: z.number() }).describe('Ray origin in world coordinates'),
  direction: z.object({ x: z.number(), y: z.number(), z: z.number() }).describe('Ray direction (will be normalized)'),
  maxDistance: z.number().positive().default(Infinity).describe('Maximum distance of the ray'),
  layerMask: z.string().optional().describe('Layer mask to filter colliders (e.g., "Default,Player")'),
  queryTriggerInteraction: QueryTriggerInteractionSchema.default('UseGlobal').describe('Whether to hit trigger colliders'),
  returnAllHits: z.boolean().default(false).describe('Return all hits instead of just the first'),
  drawDebugRay: z.boolean().default(false).describe('Draw debug ray in Scene view'),
  debugRayColor: z.string().default('red').describe('Color of debug ray'),
  debugRayDuration: z.number().default(2).describe('Duration to show debug ray in seconds')
});

@Tool({
  name: 'raycast',
  description: 'Perform raycasting to detect colliders along a ray, useful for line-of-sight, shooting, and ground detection',
  category: 'physics',
  version: '1.0.0'
})
@Tags(['unity', 'physics', 'raycast', 'detection', 'collision'])
@Examples([
  {
    description: 'Raycast downward for ground detection',
    args: {
      origin: { x: 0, y: 10, z: 0 },
      direction: { x: 0, y: -1, z: 0 },
      maxDistance: 20,
      drawDebugRay: true
    }
  },
  {
    description: 'Raycast forward from player position',
    args: {
      origin: { x: 0, y: 1, z: 0 },
      direction: { x: 0, y: 0, z: 1 },
      maxDistance: 100,
      layerMask: 'Default,Enemy',
      drawDebugRay: true,
      debugRayColor: 'green'
    }
  }
])
export class RaycastTool extends BaseTool {
  get name() { return 'raycast'; }
  get description() { return 'Perform raycasting to detect colliders along a ray, useful for line-of-sight, shooting, and ground detection'; }
  get inputSchema() { return RaycastArgsSchema; }
  get category() { return 'physics'; }

  protected async beforeExecute(args: any): Promise<void> {
    const dirMagnitude = Math.sqrt(
      args.direction.x * args.direction.x + 
      args.direction.y * args.direction.y + 
      args.direction.z * args.direction.z
    );
    if (dirMagnitude === 0) {
      throw new Error('Direction vector cannot be zero');
    }
    if (args.maxDistance <= 0) {
      throw new Error('Max distance must be positive');
    }
  }

  protected formatSuccessResponse(result: any): CallToolResult {
    const { hit, hits, origin, direction, maxDistance } = result;
    
    if (!hit && (!hits || hits.length === 0)) {
      return {
        content: [{
          type: 'text',
          text: `🔍 Raycast completed - No hits\n📍 Origin: (${origin.x}, ${origin.y}, ${origin.z})\n➡️  Direction: (${direction.x}, ${direction.y}, ${direction.z})\n📏 Max Distance: ${maxDistance}`
        }]
      };
    }

    let output = `✅ Raycast hit detected!\n📍 Origin: (${origin.x}, ${origin.y}, ${origin.z})\n➡️  Direction: (${direction.x}, ${direction.y}, ${direction.z})\n\n`;

    if (hits && hits.length > 0) {
      output += `🎯 Found ${hits.length} hit(s):\n\n`;
      hits.forEach((h: any, index: number) => {
        output += `${index + 1}. ${h.colliderName || 'Unknown'}\n`;
        output += `   • Distance: ${h.distance.toFixed(2)}\n`;
        output += `   • Point: (${h.point.x.toFixed(2)}, ${h.point.y.toFixed(2)}, ${h.point.z.toFixed(2)})\n`;
        if (h.gameObjectPath) output += `   • GameObject: ${h.gameObjectPath}\n`;
        output += '\n';
      });
    } else if (hit) {
      output += `🎯 Hit: ${hit.colliderName || 'Unknown'}\n`;
      output += `📏 Distance: ${hit.distance.toFixed(2)}\n`;
      output += `📍 Point: (${hit.point.x.toFixed(2)}, ${hit.point.y.toFixed(2)}, ${hit.point.z.toFixed(2)})\n`;
      if (hit.gameObjectPath) output += `🎮 GameObject: ${hit.gameObjectPath}\n`;
    }

    return { content: [{ type: 'text', text: output }] };
  }

  protected formatErrorResponse(error: any): CallToolResult {
    const errorMessage = error.message || 'Unknown error occurred';
    let helpText = '';
    if (errorMessage.includes('zero')) {
      helpText = '\n\n💡 Tip: Direction vector cannot be zero. Provide a valid direction.';
    }
    return {
      content: [{ type: 'text', text: `❌ Error performing raycast: ${errorMessage}${helpText}` }],
      isError: true
    };
  }
}

File: Editor/Tools/Physics/RaycastTool.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;
using McpUnity.Utils;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Perform raycasting to detect colliders along a ray, useful for line-of-sight, shooting, and ground detection
    /// Unity API: https://docs.unity3d.com/ScriptReference/Physics.Raycast.html
    /// </summary>
    public class RaycastTool : McpToolBase
    {
        public RaycastTool()
        {
            Name = "raycast";
            Description = "Perform raycasting to detect colliders along a ray, useful for line-of-sight, shooting, and ground detection";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                // Parse origin
                var originObj = parameters["origin"];
                if (originObj == null)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "origin is required", "validation_error");
                }

                Vector3 origin = new Vector3(
                    originObj["x"]?.ToObject<float>() ?? 0,
                    originObj["y"]?.ToObject<float>() ?? 0,
                    originObj["z"]?.ToObject<float>() ?? 0
                );

                // Parse direction
                var directionObj = parameters["direction"];
                if (directionObj == null)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "direction is required", "validation_error");
                }

                Vector3 direction = new Vector3(
                    directionObj["x"]?.ToObject<float>() ?? 0,
                    directionObj["y"]?.ToObject<float>() ?? -1,
                    directionObj["z"]?.ToObject<float>() ?? 0
                ).normalized;

                float maxDistance = parameters["maxDistance"]?.ToObject<float>() ?? Mathf.Infinity;
                bool returnAllHits = parameters["returnAllHits"]?.ToObject<bool>() ?? false;
                bool drawDebugRay = parameters["drawDebugRay"]?.ToObject<bool>() ?? false;
                string debugRayColor = parameters["debugRayColor"]?.ToString() ?? "red";
                float debugRayDuration = parameters["debugRayDuration"]?.ToObject<float>() ?? 2f;

                // Parse layer mask
                int layerMask = -1;
                if (parameters["layerMask"] != null)
                {
                    string layerMaskStr = parameters["layerMask"].ToString();
                    layerMask = ParseLayerMask(layerMaskStr);
                }

                // Parse query trigger interaction
                QueryTriggerInteraction queryTriggerInteraction = QueryTriggerInteraction.UseGlobal;
                if (parameters["queryTriggerInteraction"] != null)
                {
                    string qti = parameters["queryTriggerInteraction"].ToString();
                    queryTriggerInteraction = ParseQueryTriggerInteraction(qti);
                }

                JArray hitsArray = new JArray();
                int hitCount = 0;

                if (returnAllHits)
                {
                    RaycastHit[] hits = Physics.RaycastAll(origin, direction, maxDistance, layerMask, queryTriggerInteraction);
                    hitCount = hits.Length;

                    foreach (var hit in hits)
                    {
                        hitsArray.Add(CreateHitObject(hit));
                    }
                }
                else
                {
                    RaycastHit hit;
                    if (Physics.Raycast(origin, direction, out hit, maxDistance, layerMask, queryTriggerInteraction))
                    {
                        hitCount = 1;
                        hitsArray.Add(CreateHitObject(hit));
                    }
                }

                // Draw debug ray
                if (drawDebugRay)
                {
                    Color color = ParseColor(debugRayColor);
                    Vector3 endPoint = hitCount > 0 ? origin + direction * ((JObject)hitsArray[0])["distance"].ToObject<float>() : origin + direction * maxDistance;
                    Debug.DrawLine(origin, endPoint, color, debugRayDuration);
                }

                return new JObject
                {
                    ["success"] = true,
                    ["hitCount"] = hitCount,
                    ["hits"] = hitsArray,
                    ["origin"] = new JObject { ["x"] = origin.x, ["y"] = origin.y, ["z"] = origin.z },
                    ["direction"] = new JObject { ["x"] = direction.x, ["y"] = direction.y, ["z"] = direction.z },
                    ["maxDistance"] = maxDistance
                };
            }
            catch (Exception ex)
            {
                McpLogger.LogError($"RaycastTool error: {ex.Message}");
                return McpUnitySocketHandler.CreateErrorResponse($"Failed: {ex.Message}", "execution_error");
            }
        }

        private JObject CreateHitObject(RaycastHit hit)
        {
            return new JObject
            {
                ["collider"] = hit.collider.name,
                ["distance"] = hit.distance,
                ["point"] = new JObject { ["x"] = hit.point.x, ["y"] = hit.point.y, ["z"] = hit.point.z },
                ["normal"] = new JObject { ["x"] = hit.normal.x, ["y"] = hit.normal.y, ["z"] = hit.normal.z },
                ["gameObject"] = hit.collider.gameObject.name,
                ["tag"] = hit.collider.tag,
                ["layer"] = LayerMask.LayerToName(hit.collider.gameObject.layer)
            };
        }

        private int ParseLayerMask(string layerMaskStr)
        {
            if (string.IsNullOrEmpty(layerMaskStr))
                return -1;

            string[] layers = layerMaskStr.Split(',');
            int mask = 0;
            foreach (string layer in layers)
            {
                int layerIndex = LayerMask.NameToLayer(layer.Trim());
                if (layerIndex >= 0)
                {
                    mask |= (1 << layerIndex);
                }
            }
            return mask == 0 ? -1 : mask;
        }

        private QueryTriggerInteraction ParseQueryTriggerInteraction(string qti)
        {
            switch (qti)
            {
                case "UseGlobal": return QueryTriggerInteraction.UseGlobal;
                case "Ignore": return QueryTriggerInteraction.Ignore;
                case "Collide": return QueryTriggerInteraction.Collide;
                default: return QueryTriggerInteraction.UseGlobal;
            }
        }

        private Color ParseColor(string colorStr)
        {
            switch (colorStr.ToLower())
            {
                case "red": return Color.red;
                case "green": return Color.green;
                case "blue": return Color.blue;
                case "yellow": return Color.yellow;
                case "white": return Color.white;
                case "black": return Color.black;
                default: return Color.red;
            }
        }
    }
}
